src/include/distributed/shard_cleaner.h

```diff
@@ -17,8 +17,98 @@ extern bool DeferShardDeleteOnMove;
 extern double DesiredPercentFreeAfterMove;
 extern bool CheckAvailableSpaceBeforeMove;
 
+extern bool DeferShardDeleteOnSplit;
+extern int NextOperationId;
+extern int NextCleanupRecordId;
+
 extern int TryDropOrphanedShards(bool waitForLocks);
-extern int DropOrphanedShards(bool waitForLocks);
 extern void DropOrphanedShardsInSeparateTransaction(void);
 
+/* Members for cleanup infrastructure */
+typedef uint64 OperationId;
+extern OperationId CurrentOperationId;
+
+/*
+ * CleanupResource represents the Resource type in cleanup records.
+ */
+typedef enum CleanupObject
+{
+	CLEANUP_OBJECT_INVALID = 0,
+	CLEANUP_OBJECT_SHARD_PLACEMENT = 1
+} CleanupObject;
+
+/*
+ * CleanupPolicy represents the policy type for cleanup records.
+ */
+typedef enum CleanupPolicy
+{
+	/*
+	 * Resources that are transient and always need clean up after the operation is completed.
+	 * (Example: Dummy Shards for Non-Blocking splits)
+	 */
+	CLEANUP_ALWAYS = 0,
+
+	/*
+	 * Resources that are cleanup only on failure.
+	 * (Example: Split Children for Blocking/Non-Blocking splits)
+	 */
+	CLEANUP_ON_FAILURE = 1,
+
+	/*
+	 * Resources that need 'deferred' clean up only on success .
+	 * (Example: Parent child being split for Blocking/Non-Blocking splits)
+	 */
+	CLEANUP_DEFERRED_ON_SUCCESS = 2,
+} CleanupPolicy;
+
+/* Global Constants */
+#define INVALID_OPERATION_ID 0
+#define INVALID_CLEANUP_RECORD_ID 0
+
+/* APIs for cleanup infrastructure */
+
+/*
+ * RegisterNewOperationNeedingCleanup is be called by an operation to register
+ * for cleanup.
+ */
+extern OperationId RegisterNewOperationNeedingCleanup(void);
+
+/*
+ * InsertCleanupRecordInCurrentTransaction inserts a new pg_dist_cleanup entry
+ * as part of the current transaction.
+ *
+ * This is primarily useful for deferred cleanup (CLEANUP_DEFERRED_ON_SUCCESS)
+ * scenarios, since the records would roll back in case of failure.
+ */
+extern void InsertCleanupRecordInCurrentTransaction(CleanupObject objectType,
+													char *objectName,
+													int nodeGroupId,
+													CleanupPolicy policy);
+
+/*
+ * InsertCleanupRecordInSeparateTransaction inserts a new pg_dist_cleanup entry
+ * in a separate transaction to ensure the record persists after rollback.
+ *
+ * This is used in scenarios where we need to cleanup resources on operation
+ * completion (CLEANUP_ALWAYS) or on failure (CLEANUP_ON_FAILURE).
+ */
+extern void InsertCleanupRecordInSubtransaction(CleanupObject objectType,
+												char *objectName,
+												int nodeGroupId,
+												CleanupPolicy policy);
+
+/*
+ * FinalizeNewOperationNeedingCleanupOnFailure is be called by an operation to signal
+ * completion on failure. This will trigger cleanup of appropriate resources
+ * and cleanup records.
+ */
+extern void FinalizeNewOperationNeedingCleanupOnFailure(void);
```

Comment:
NewOperation -> Operation?

The "New" was because of "start new", though I think we can also RegisterNewOperation -> RegisterNewOperation

Reply:
Updated